SRV record resource and datasource

bodgit opened the pull request:
This builds on #69 and adds SRV record support.

Records can be created with the following:
```hcl
resource "dns_srv_record_set" "sip" {
  zone = "example.com."
  name = "_sip._tcp"
  ttl  = 86400

  srv {
    priority = 10
    weight   = 60
    target   = "sip.example.com."
    port     = 5060
  }
}
```
Records can be queried with:
```hcl
data "dns_srv_record_set" "kerberos" {
  service = "_kerberos._tcp.example.com."
}

output "kerberos" {
  value = "${data.dns_srv_record_set.kerberos.srv.0.target}"
}
```
The only problem I have with the data source is that I don't know of a reliable external SRV record I can query for tests, so they're currently commented out.

Fixes #45

katbyte commented:
👋 @bodgit 

I'll take a look at this one once master has been merged back in

bodgit commented:
I'll fix up the conflict now the MX PR has been merged. In the meantime, the outstanding issue is having a test SRV record to query in the datasource tests. Is there a Hashicorp test zone or something similar that could host one?

jadunham1 commented:
What is the status on this request?  Would be really nice to have SRV records for our use cases.

trevor-viljoen commented:
@bodgit @katbyte Is this ready to merge or are we still waiting on that test to have access to a live `SRV` record?

bodgit commented:
@trevor-viljoen I would say it's still waiting on the tests working. I doubt it will be merged without that, which is understandable. Having said that I can't do much without feedback from someone at Hashicorp.

iinuwa commented:
Hi I'm new here. Thanks for the work on this!
Maybe we could [contact IANA](https://www.iana.org/contact) about adding a test SRV record to the example.(org|com|net) domains? That may be helpful for the broader community.

trevor-viljoen commented:
It would be nice if Hashi would at least acknowledge this PR. It would be useful in standing up Open Shift in a lab environment as referenced above and is the reason I came looking to see if this was possible through terraform.

DanyC97 commented:
@katbyte @mildwonkey @apparentlymart anyone of you able to help reviewing this PR please ? it will help a lot if the work started by @bodgit gets merged, thanks for consideration

DanyC97 commented:
@pselle any chance you can help out finding someone internally who can review this PR? thanks in advance

appilon commented:
Hello, I will do my best to help get this PR addressed and merged.

appilon commented:
This looks very interesting. I am quite rusty with DNS work (so please bear with me, in particular I have never worked with SRV records). Some initial thoughts are the names of the fields.
I see the consistent use of `zone` and `name` as fields, similar to the other record types. I do appreciate the consistency, however the [RFC](https://tools.ietf.org/html/rfc2782) breaks the address into 3 parts, `Service`, `Proto`, and `Name`, where `Name` is mapping to what this PR defines as `zone`. I'm only contemplating which convention to follow.

As for a test record for the datasource, I quickly found an online tool that we can refer to its own [HTTP/TCP service/protocol](https://mxtoolbox.com/SuperTool.aspx?action=srv%3a_http._tcp.mxtoolbox.com&run=toolpage). In the (maybe near) future we might invest in some infrastructure for these acceptance tests, but we will get by for now on whatever public examples we can find.

appilon commented:
After reflection and discussion it was determined we should just keep the current `zone/name` model to represent the record as its consistent with the other types and quite frankly "plays nicer" with the various DNS service providers that typically just accept a single string with the addresses  regardless of type anyways. I will push a commit getting acceptance tests for the data source and hopefully can get this merged.

appilon commented:
@bodgit I added that srv record I found to the data source test, I changed the test to just confirm the target, at first it looked like it was expecting a hash? however the datasource was setup as a list (and not a set). The tests pass so let me know what you think and we can get this merged.

bodgit commented:
> I also did not have time to read through all 34 pages of the RFC 😂and this could be a rookie question but is it allowed/make any sense to have the TTL not be the same across the whole set? My gut tells me the way it's implemented is correct/how practitioners would configure these things but just checking.

I was going to say, it will take me ages to find the text in the RFC, however I surprised myself with my own google-fu. [RFC 2181 section 5.2](https://tools.ietf.org/html/rfc2181#section-5) has this text which sounds like what I based the behaviour on:

>    Resource Records also have a time to live (TTL).  It is possible for
   the RRs in an RRSet to have different TTLs.  No uses for this have
   been found that cannot be better accomplished in other ways.  This
   can, however, cause partial replies (not marked "truncated") from a
   caching server, where the TTLs for some but not all the RRs in the
   RRSet have expired.

>   Consequently the use of differing TTLs in an RRSet is hereby
   deprecated, the TTLs of all RRs in an RRSet must be the same.

>   Should a client receive a response containing RRs from an RRSet with
   differing TTLs, it should treat this as an error.  If the RRSet
   concerned is from a non-authoritative source for this data, the
   client should simply ignore the RRSet, and if the values were
   required, seek to acquire them from an authoritative source.  Clients
   that are configured to send all queries to one, or more, particular
   servers should treat those servers as authoritative for this purpose.
   **Should an authoritative source send such a malformed RRSet, the
   client should treat the RRs for all purposes as if all TTLs in the
   RRSet had been set to the value of the lowest TTL in the RRSet.**  In
   no case may a server send an RRSet with TTLs not all equal.

bodgit commented:
> @bodgit I added that srv record I found to the data source test, I changed the test to just confirm the target, at first it looked like it was expecting a hash? however the datasource was setup as a list (and not a set). The tests pass so let me know what you think and we can get this merged.

@appilon Thanks for finding the public SRV record. I will look at the tests again and confirm.

DanyC97 commented:
@bodgit @appilon many many thanks! 💯

bodgit commented:
@appilon I've updated the test code to match what was done for the MX datasource. It now checks all fields match the expected values.

While I was there, I noticed when I ran the tests with `make test` that the TXT datasource tests were always running yet the SRV tests were not. I found the difference is down to using `resource.Test` vs `resource.UnitTest` to define the tests, the latter doesn't require `TF_ACC` being set. As these tests don't require any external API apart from just a basic working DNS resolver I updated the other datasource tests so they're now consistently using `resource.UnitTest`. If these should still be using `resource.Test` let me know.

appilon commented:
@bodgit that's all good, ready to merge all acctests/unitests pass on TeamCity

trevor-viljoen commented:
Thanks to @bodgit, @appilon, and @DanyC97 for making this happen.
